pallets/portfolio/src/lib.rs

```diff
@@ -0,0 +1,335 @@
+// General Public License for more details.
+//
+// You should have received a copy of the GNU General Public License
+// along with this program. If not, see <http://www.gnu.org/licenses/>.
+
+//! # Portfolio Module
+//!
+//! ## Overview
+//!
+//! The portfolio module provides the essential extrinsics to manage asset portfolios, public
+//! functions for integration of portfolios into other pallets, and implementations of RPC getters.
+//!
+//! ## Interface
+//!
+//! ### Dispatchable Functions
+//!
+//! - `create_portfolio`: Creates a new user portfolio.
+//! - `delete_portfolio`: Deletes an existing user portfolio.
+//! - `move_portfolio`: Moves specified amounts of assets from one portfolio to another portfolio
+//!   of the same DID.
+//! - `rename_portfolio`: Renames a user portfolio.
+//!
+//! ### Public Functions
+//!
+//! - `default_portfolio_balance`: Returns the ticker balance of the identity's default portfolio.
+//! - `user_portfolio_balance`: Returns the ticker balance of an identity's user portfolio.
+//! - `set_default_portfolio_balance`: Sets the ticker balance of the identity's default portfolio.
+//! - `rpc_get_portfolios`: An RPC function that lists all user-defined portfolio number-name pairs.
+//! - `rpc_get_portfolio_assets`: Ensures that there is no portfolio with the desired name yet.
+
+#![cfg_attr(not(feature = "std"), no_std)]
+
+use codec::{Decode, Encode};
+use core::result::Result;
+use frame_support::{
+    decl_error, decl_event, decl_module, decl_storage, dispatch::DispatchResult, ensure,
+    IterableStorageDoubleMap,
+};
+use frame_system::{self as system, ensure_signed};
+use polymesh_common_utilities::{identity::Trait as IdentityTrait, CommonTrait, Context};
+use polymesh_primitives::{IdentityId, PortfolioId, PortfolioName, PortfolioNumber, Ticker};
+use sp_arithmetic::traits::Saturating;
+use sp_std::{convert::TryFrom, prelude::Vec};
+
+type Identity<T> = pallet_identity::Module<T>;
+
+/// The ticker and balance of an asset to be moved from one portfolio to another.
+#[derive(Encode, Decode, Clone, Debug, Default, PartialEq, Eq, PartialOrd, Ord)]
+pub struct MovePortfolioItem<Balance> {
+    /// The ticker of the asset to be moved.
+    pub ticker: Ticker,
+    /// The balance of the asset to be moved.
+    pub amount: Balance,
+}
+
+pub trait Trait: CommonTrait + IdentityTrait {
+    type Event: From<Event<Self>> + Into<<Self as frame_system::Trait>::Event>;
+}
+
+decl_storage! {
+    trait Store for Module<T: Trait> as Session {
+        /// The set of existing portfolios with their names. If a certain pair of a DID and
+        /// portfolio number maps to `None` then such a portfolio doesn't exist. Conversely, if a
+        /// pair maps to `Some(name)` then such a portfolio exists and is called `name`.
+        pub Portfolios get(fn portfolios):
+            double_map hasher(blake2_128_concat) IdentityId, hasher(twox_64_concat) PortfolioNumber =>
```

Comment:
`IdentityId` should use simpler hasher like `twox_64_concat`. At least, we have that in other modules, so we should unify that.